### PR TITLE
Add luaprinter with centronics and rs232 luaprinter devices

### PR DESCRIPTION
**Pull request by goldnchild**
I've tested it with ubuntu, I don't know how it works with mac or windows.

It maintains a luaprinterlist structure and removes the item from the list in the class destructor.

With the luaprinter, you can do things like look at the byte streams:

for example, to inspect the characters in the buffer:
lp1 = manager:machine().lp[1]
a,b,c = lp1:getbuffer()
print(a,b,c)
for i=c,b do print(i+1,a[i+1]) end

It also provides utility functions for printer devices and saving printer bitmaps.

### CODE REVIEW

File: src/emu/luaprinter.h [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    Utility class for printing functions
+
+***************************************************************************/
+
+#ifndef MAME_LUAPRINTER_H
+#define MAME_LUAPRINTER_H
+
+#pragma once
+
+#include "emu.h"
```
Comment (cuavas) [inline src/emu/luaprinter.h:15]: You shouldn’t include emu.h from headers like this.

File: src/emu/luaprinter.h [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+#include "png.h"
+#include "emuopts.h"
+
+class luaprinter {
+
+public:
+	luaprinter(device_t& thisdevice, std::string dummyname);
+	~luaprinter() { lp_removefromprinterlist(); }
```
Comment (cuavas) [inline src/emu/luaprinter.h:23]: If this is supposed to be a mixin that integrates with the `device_t` lifecycle, it should implement `device_interface`.  Also, passing a string by value like this is usually not what you want to do.

File: src/emu/luaprinter.h [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+	device_t *m_lp_mydevice;
+	static std::vector<luaprinter *> luaprinterlist;
```
Comment (cuavas) [inline src/emu/luaprinter.h:26]: You shouldn’t be doing this.  Use an iterator to walk the device graph and find devices that derive from your mixin.

File: src/emu/luaprinter.h [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    int getpixel(int x, int y) { return m_lp_bitmap->pix32(y,x); };
+
+    void setheadpos(int x, int y) { m_lp_xpos=x;m_lp_ypos=y; };
+    std::tuple<int,int>getheadpos() { return std::make_tuple(m_lp_xpos,m_lp_ypos); };
+	int pagewidth() { return m_lp_bitmap->width(); }
+	int pageheight() { return m_lp_bitmap->height(); }
+	std::tuple<int,int> pagesize() { return std::make_tuple(m_lp_bitmap->width(), m_lp_bitmap->height()); }
+
+    std::tuple<std::array<unsigned char,BUFFERSIZE>&,int,int>  getbuffer() {
+		return std::make_tuple(std::ref(m_lp_printerbuffer),m_lp_head,m_lp_tail); };
```
Comment (cuavas) [inline src/emu/luaprinter.h:85]: Exposing implementation details like buffer pointers is generally not a good idea.

File: src/emu/luaprinter.h [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	int m_lp_printheadxsize = 15;
+	int m_lp_printheadysize = 30;
+	int m_lp_distfrombottom = 50;
+	int m_lp_clearlinepos = 0;
+	int m_lp_papercolor=0xffffff;
+	int m_lp_pagedirty = 0;
+    bitmap_rgb32 *m_lp_bitmap;  // pointer to bitmap
+    std::string m_lp_luaprintername;
+    std::string m_lp_snapshotdir;
+	static time_t m_lp_session_time;
```
Comment (cuavas) [inline src/emu/luaprinter.h:47]: Having a static member for this doesn’t strike me as a good idea.

**Review by cuavas — CHANGES_REQUESTED** [review]
As well as the specific issues commented in-place, the code is borderline unreadable with a mixture of spaces and tabs for indentation, inconsistent indentation and spacing, and no organisation at all in the `luaprinter` class.

Please reformat the code in Allman style with one tab indent per scope level.

Also, organise the `luaprinter` class properly, using the `public`, `protected` and `private` visibility levels appropriately.

I think there are a lot of other design issues with this, but with the code in its current state it’s too much of a mess to understand.

**Comment by goldnchild** [comment]
ok, I'll try to figure out how mix in devices work

**Comment by goldnchild** [comment]
Thanks for looking at it, I'll work on it some more.

**Pull request closed without merging**